File: frontend/src/types/channel.ts
/**
 * Channel type definitions
 * Channels organize multiple devices together (similar to ThingSpeak channels)
 */

export interface Channel {
  id: string;
  tenant_id: string;
  name: string;
  description?: string;
  icon?: string; // emoji or icon identifier
  color?: string; // color theme
  metadata?: Record<string, any>;
  created_by?: string;
  created_at: string;
  updated_at: string;

  // Computed fields (from joins)
  device_count?: number;
  online_count?: number;
}

export interface ChannelCreate {
  tenant_id: string;
  name: string;
  description?: string;
  icon?: string;
  color?: string;
  metadata?: Record<string, any>;
}

export interface ChannelUpdate {
  name?: string;
  description?: string;
  icon?: string;
  color?: string;
  metadata?: Record<string, any>;
}

export interface ChannelListResponse {
  channels: Channel[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface ChannelFilters {
  tenant_id: string;
  search?: string;
  page?: number;
  page_size?: number;
}

// Predefined color themes for channels
export const CHANNEL_COLORS = {
  green: { bg: 'bg-green-50', border: 'border-green-200', text: 'text-green-800', badge: 'bg-green-100' },
  blue: { bg: 'bg-blue-50', border: 'border-blue-200', text: 'text-blue-800', badge: 'bg-blue-100' },
  purple: { bg: 'bg-purple-50', border: 'border-purple-200', text: 'text-purple-800', badge: 'bg-purple-100' },
  orange: { bg: 'bg-orange-50', border: 'border-orange-200', text: 'text-orange-800', badge: 'bg-orange-100' },
  red: { bg: 'bg-red-50', border: 'border-red-200', text: 'text-red-800', badge: 'bg-red-100' },
  yellow: { bg: 'bg-yellow-50', border: 'border-yellow-200', text: 'text-yellow-800', badge: 'bg-yellow-100' },
  indigo: { bg: 'bg-indigo-50', border: 'border-indigo-200', text: 'text-indigo-800', badge: 'bg-indigo-100' },
  pink: { bg: 'bg-pink-50', border: 'border-pink-200', text: 'text-pink-800', badge: 'bg-pink-100' },
  gray: { bg: 'bg-gray-50', border: 'border-gray-200', text: 'text-gray-800', badge: 'bg-gray-100' },
} as const;

export type ChannelColor = keyof typeof CHANNEL_COLORS;

// Default icons for channels
export const DEFAULT_CHANNEL_ICONS = [
  '🌾', // Farm
  '💻', // Server Room
  '🏢', // Building
  '🏭', // Manufacturing
  '🏠', // Home
  '🏥', // Healthcare
  '🏫', // Education
  '🏪', // Retail
  '🚗', // Transportation
  '⚡', // Energy
  '💧', // Water
  '🌡️', // Temperature
  '📦', // Uncategorized
] as const;

File: backend/app/models/channel.py
"""
Channel Models
For organizing devices into channels (similar to ThingSpeak channels)
"""

from pydantic import BaseModel, Field, field_validator
from typing import Optional, Dict, Any
from datetime import datetime
from uuid import UUID


class ChannelBase(BaseModel):
    """Base channel model with common fields"""
    name: str = Field(..., min_length=1, max_length=255, description="Channel name")
    description: Optional[str] = Field(None, description="Channel description")
    icon: Optional[str] = Field(None, max_length=50, description="Emoji or icon identifier")
    color: Optional[str] = Field(None, max_length=50, description="Color theme")
    metadata: Optional[Dict[str, Any]] = Field(default_factory=dict, description="Additional metadata")


class ChannelCreate(ChannelBase):
    """Model for creating a new channel"""
    tenant_id: UUID = Field(..., description="Tenant ID for multi-tenancy")


class ChannelUpdate(BaseModel):
    """Model for updating a channel"""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    icon: Optional[str] = Field(None, max_length=50)
    color: Optional[str] = Field(None, max_length=50)
    metadata: Optional[Dict[str, Any]] = None


class Channel(ChannelBase):
    """Complete channel model with all fields"""
    id: UUID
    tenant_id: UUID
    created_by: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime

    # Computed fields (from queries)
    device_count: Optional[int] = Field(None, description="Number of devices in this channel")
    online_count: Optional[int] = Field(None, description="Number of online devices")

    model_config = {
        "from_attributes": True,
        "json_schema_extra": {
            "example": {
                "id": "70000000-0000-0000-0000-000000000001",
                "tenant_id": "10000000-0000-0000-0000-000000000001",
                "name": "Smart Farm",
                "description": "Agricultural IoT devices",
                "icon": "🌾",
                "color": "green",
                "metadata": {},
                "device_count": 5,
                "online_count": 3,
                "created_at": "2024-01-01T00:00:00Z",
                "updated_at": "2024-01-01T00:00:00Z"
            }
        }
    }


class ChannelListResponse(BaseModel):
    """Response model for list of channels"""
    channels: list[Channel]
    total: int
    page: int
    page_size: int
    total_pages: int
